SourceCode/HW6+/HW6+/Math3D.hpp

```diff
@@ -0,0 +1,335 @@
+#pragma once
+
+#include <cstdint>
+#include <numbers>
+#include <string>
+#include <cmath>
+
+constexpr float DegToRad(float degrees)
+{
+	return degrees * (std::numbers::pi_v<float> / 180.f);
+}
+
+constexpr float RadToDeg(float radians)
+{
+	return radians * (180.f / std::numbers::pi_v<float>);
+}
+
+struct RGB 
+{
+	uint8_t r;
+	uint8_t g;
+	uint8_t b;
+
+	std::string ToString() const
+	{
+		return std::to_string(r) + " " + std::to_string(g) + " " + std::to_string(b);
+	}
+};
+
+struct Vector3
+{
+	float x;
+	float y;
+	float z;
+
+	Vector3(float x, float y, float z) : x(x), y(y), z(z) {}
+
+	Vector3& operator *=(float s)
+	{
+		x *= s;
+		y *= s;
+		z *= s;
+		return (*this);
+	}
+
+	Vector3& operator /=(float s)
+	{
+		s = 1.f / s;
+		x *= s;
+		y *= s;
+		z *= s;
+		return (*this);
+	}
+
+	Vector3& operator +=(const Vector3& v)
+	{
+		x += v.x;
+		y += v.y;
+		z += v.z;
+		return (*this);
+	}
+
+	Vector3& operator -=(const Vector3& v)
+	{
+		x -= v.x;
+		y -= v.y;
+		z -= v.z;
+		return (*this);
+	}
+
+	float Magnitude() const
+	{
+		return std::sqrt(x * x + y * y + z * z);
+	}
+
+	RGB ToRGB() const
+	{
+		return RGB{ static_cast<uint8_t>(x * 256), static_cast<uint8_t>(y * 256), static_cast<uint8_t>(z * 256) };
+	}
+
+	std::string ToString() const
+	{
+		return "(" + std::to_string(x) + ", " + std::to_string(y) + ", " + std::to_string(z) + ")";
+	}
+};
+
+inline Vector3 operator +(const Vector3& a, const Vector3& b)
+{
+	return {a.x + b.x, a.y + b.y, a.z + b.z};
+}
+
+inline Vector3 operator -(const Vector3& a, const Vector3& b)
+{
+	return {a.x - b.x, a.y - b.y, a.z - b.z};
+}
+
+inline Vector3 operator *(const Vector3& v, float s)
+{
+	return { v.x * s, v.y * s, v.z * s };
+}
+
+inline Vector3 operator /(const Vector3& v, float s)
+{
+	s = 1.f / s;
+	return { v.x * s, v.y * s, v.z * s };
+}
+
+inline Vector3 Normalize(const Vector3& v)
+{
+	return v / v.Magnitude();
+}
+
+inline Vector3 Cross(const Vector3& a, const Vector3& b)
+{
+	return {
+		a.y * b.z - a.z * b.y,
+		a.z * b.x - a.x * b.z,
+		a.x * b.y - a.y * b.x
+	};
+}
+
+inline float Dot(const Vector3& a, const Vector3& b)
+{
+	return a.x * b.x + a.y * b.y + a.z * b.z;
+}
+
+struct Point3 : Vector3
+{
+	Point3() = default;
+
+	Point3(float a, float b, float c) : Vector3(a, b, c) {}
+
+	Point3& operator =(const Vector3& v)
+	{
+		x = v.x;
+		y = v.y;
+		z = v.z;
+		return (*this);
+	}
+};
+
+inline Point3 operator +(const Point3& a, const Vector3& b)
+{
+	return Point3(a.x + b.x, a.y + b.y, a.z + b.z);
+}
+
+inline Point3 operator -(const Point3& a, const Vector3& b)
+{
+	return Point3(a.x - b.x, a.y - b.y, a.z - b.z);
+}
+
+inline Vector3 operator -(const Point3& a, const Point3& b)
+{
+	return Vector3(a.x - b.x, a.y - b.y, a.z - b.z);
+}
+
+struct Ray
+{
+	Vector3 origin;
+	Vector3 directionN;
+
+	Vector3 operator()(float t) const
+	{
+		return origin + directionN * t;
+	}
```

Comment:
And since this is a point along the ray, it would again be a point, not a vector.

All of those are just semantics of course, and I would probably just use the same class for Point and Vector.